### PR TITLE
ICU-22262 Automate environment test

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22262
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [ ] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

Configure 17+21 = 38 jobs to run icu tests under
500+ locales with LC_ALL= environment variable, and
600+ timezone with TZ= environment variable.

Fix broken tests discovered:
1. Ignore error in cintltst/cdtrgtst.c if the calendar of the default
   locale is not "gregorian" because the expectation is reasonable
   for 'gregorian' calendar only.
2. Consider the string are the same if the numerical value of the digit
   are the same for tests inside cintltst/cdtrgtst.c.
3. Skip the TestCDefaultLocale() error in cintltst/cloctst.c
   if LC_ALL env is set.
4. Fix the test which incorrectly assume the string iteration is not end
   (under "wo" locale") and avoid incorrect expectation in TestAssignment()
   of intltest/itercoll.cpp (see [ICU-22273](https://unicode-org.atlassian.net/browse/ICU-22273) )
5. Skip Test4139940() error in intltest/loctest.cpp
   if the calendar is not "gregorian".
6. Log known issue [ICU-22274](https://unicode-org.atlassian.net/browse/ICU-22274) for failure of detectHostTimeZone()'s raw offset
   != host timezone's offset while the default TimeZones is one of the
   following troublesome TimeZones:
   "Africa/Casablanca", "Europe/Dublin", "America/Godthab", "America/Nuuk",
   or "Africa/El_Aaiun".

Following
https://unicode-org.github.io/icu/processes/release/tasks/integration.html#run-the-environment-tests

[ICU-22273]: https://unicode-org.atlassian.net/browse/ICU-22273?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by FrankYFTang** [comment]
with 30 tests (locales or timezone) per job (eg shard), each job take about 36-45 minutes to run, faster than two othe pre-existing ci . 
My intention is to land this PR as is, and after the landing, create another PR to comment out (not remove, but just comment out) the two lines
```
  pull_request:
    branches: '**'
```
under the line "For BRS" on the top of file .github/workflows/icu_envtest.yml
so Env Test will stop running on every pull_request but only run by cron on every Sat 10:23 Sat UTC (which is about every Friday night PST) because we do not need to run this so frequently and Weekly run should be good enough. 
In the future if people want to run it on demand, just uncomment that two line and push a PR and such push (no need to merge, just push) to github will trigger gihub to run the Env tests.

**Comment by FrankYFTang** [comment]
currently, the env test set up (30 tests per job) require less time (which is about 45mins) to run than
uconfig_variation-check-unit-tests and uconfig_variation-check-all-header-tests (which both took about 55mins)
If we speed up these two tests, we may want to reconfigure the number of tests per job in the all env test to make it complete faster.

**Comment by markusicu** [comment]
> My intention is to land this PR as is, and after the landing, create another PR to comment out (not remove, but just comment out) the two lines ... so Env Test will stop running on every pull_request but only run by cron ... Weekly run should be good enough. In the future if people want to run it on demand, just uncomment that two line and push a PR and such push (no need to merge, just push) to github will trigger gihub to run the Env tests.

@mihnita found that it's possible to add a few config lines that define a command line interface to manually trigger a GitHub Action, no pending PR needed. That seems better.

File: .github/workflows/icu_envtest.yml
Comment (markusicu) [inline .github/workflows/icu_envtest.yml:17]: Please ask @mihnita for instructions for setting up a command line interface for a manual trigger.

File: .github/workflows/icu_envtest.yml
Comment (markusicu) [inline .github/workflows/icu_envtest.yml:32]:
```suggestion
        # 17x30 = 510 > 502 (the number of locales in 'locale -a').
```

File: icu4c/source/test/cintltst/cdtrgtst.c
Comment (markusicu) [inline icu4c/source/test/cintltst/cdtrgtst.c:264]: This passes if the characters differ and neither is a digit. -->

```suggestion
        if (ch1 != ch2) {
            int32_t digit1 = u_charDigitValue(ch1);
            int32_t digit2 = u_charDigitValue(ch2);
            if (digit1 < 0 || digit1 != digit2) { return false; }
        }
```

File: icu4c/source/test/cintltst/cdtrgtst.c
Comment (markusicu) [inline icu4c/source/test/cintltst/cdtrgtst.c:298]:
```suggestion
          log_verbose("Skipping Test4059917 when the default date time format is not using gregorian calendar.");
```

File: .github/workflows/icu_envtest.yml
Comment (FrankYFTang) [reply .github/workflows/icu_envtest.yml]: @mihnita Is workflow_dispatch stated in  https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch what you are talking about?

**Comment by FrankYFTang** [comment]
Ready for Review

File: .github/workflows/icu_envtest.yml
Comment (FrankYFTang) [reply .github/workflows/icu_envtest.yml]: it is hard to debug how to set up that manual trigger without this yaml file in the repo. Could we resolve that part later? I put in workflow_dispatch but I am not 100% sure it work unless we land this. (not sure would that only show up after the config is landing or not)

Comment (mihnita) [reply .github/workflows/icu_envtest.yml]: Yes, it is `workflow_dispatch`

> it is hard to debug how to set up that manual trigger without this yaml file in the repo

I was able to play with it in a fork + branch. I can help resolve it later, or in this PR, your call.

File: .github/workflows/icu_envtest.yml
Comment (mihnita) [inline .github/workflows/icu_envtest.yml:9]: Yes, this should be enough for manual trigger

File: .github/workflows/icu_envtest.yml
Comment (echeran) [reply .github/workflows/icu_envtest.yml]: FYI, after configuring `workflow_dispatch`, you are not limited to the CLI for a manual trigger. You can also use the UI, which seems much less cumbersome and error-prone: https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow?tool=webui

File: icu4c/source/test/cintltst/cdtrgtst.c
Comment (echeran) [inline icu4c/source/test/cintltst/cdtrgtst.c:252]:
```suggestion
// numerical values are the same if we ignore the numbering system.
```

File: .github/workflows/icu_envtest.yml
Comment (FrankYFTang) [reply .github/workflows/icu_envtest.yml:9]: how to trigger manually?

**Comment by FrankYFTang** [comment]
PTAL

**Comment by FrankYFTang** [comment]
ping

**Review by markusicu — APPROVED** [review]
lgtm tnx!

As a follow-up after merging this PR, please edit the BRS task description and add near its top a link to this .yml file and say that it runs periodically and can also be triggered manually / on demand.

**Comment by FrankYFTang** [comment]
the manual trigger up show up in https://github.com/unicode-org/icu/actions/workflows/icu_envtest.yml after I landed the PR

**Comment by markusicu** [comment]
> the manual trigger up show up in https://github.com/unicode-org/icu/actions/workflows/icu_envtest.yml after I landed the PR

Very nice! :-)

**Comment by markusicu** [comment]
> lgtm tnx!
> 
> As a follow-up after merging this PR, please edit the BRS task description and add near its top a link to this .yml file and say that it runs periodically and can also be triggered manually / on demand.

At https://github.com/unicode-org/icu/blob/main/docs/processes/release/tasks/integration.md?plain=1#L399
after the “Run the environment tests” heading, please add something like

> Since ICU 73, this test has been automated via a [GitHub Action](https://github.com/unicode-org/icu/actions/workflows/icu_envtest.yml) which runs periodically. You can also trigger it manually from that UI page.

**Pull request merged**
